src/tls.c

```diff
@@ -3412,20 +3453,71 @@ static int TLSX_CSR_Parse(WOLFSSL* ssl, const byte* input, word16 length,
                 return ret;
             }
             FreeDecodedCert(cert);
-            XFREE(cert, ssl->heap, DYNAMIC_TYPE_DCERT);
-
             extension = TLSX_Find(ssl->extensions, TLSX_STATUS_REQUEST);
             csr = extension ?
                 (CertificateStatusRequest*)extension->data : NULL;
             if (csr == NULL)
                 return MEMORY_ERROR;
 
-            request = &csr->request.ocsp;
-            ret = CreateOcspResponse(ssl, &request, &csr->response);
+            request = &csr->request.ocsp[0];
+            ret = CreateOcspResponse(ssl, &request, &csr->responses[0]);
+            if (request != &csr->request.ocsp[0] &&
+                    ssl->buffers.weOwnCert) {
+                /* request will be allocated in CreateOcspResponse() */
+                FreeOcspRequest(request);
+                XFREE(request, ssl->heap, DYNAMIC_TYPE_OCSP_REQUEST);
+            }
             if (ret != 0)
                 return ret;
-            if (csr->response.buffer)
+            if (csr->responses[0].buffer)
                 TLSX_SetResponse(ssl, TLSX_STATUS_REQUEST);
+
+            {
```

Comment:
Move this out into its own function as it appears self-contained.

Reply:
Fixed

Reply:
I don't see a change here.

Reply:
I misunderstood. Move the block to function.

I realize the block bypasses error code from CheckOcspRequest(). This causes existing ocsp-stapling.test has them passed without any modification.

Do you think we should keep existing behavior? To introduce the changes, we should have a new macro to enable it?

Reply:
I add new macro to enable this behavior to maintain existing one.